Add Spandrel Force Result

GCRA101 opened the pull request:
<!-- PLEASE ENSURE YOU REVIEW THE CONTENT OF EACH PR CAREFULLY, INCLUDING SUBSEQUENT COMMENTS BY YOURSELF OR OTHERS. -->
<!-- IN PARTICULAR PLEASE ENSURE THAT SENSITIVE OR INAPPROPRIATE INFORMATION IS NOT UPLOADED -->

 ### Issues addressed by this PR
<!-- Add reference(s) to issue(s) solved by this PR. Please use keyword Fixes/Closes as per https://help.github.com/articles/closing-issues-using-keywords/ -->

 Closes #430 

 <!-- Add short description of what has been fixed -->


 ### Test files
Grasshopper File
[https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23457-AddSpandrelForceResult/Test%20Script.gh?csf=1&web=1&e=caphlX](url)
ETABS File
[https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/02_Pull%20Request/BHoM/ETABS_Toolkit/%23457-AddSpandrelForceResult/Test%20ETABS%20Model.EDB?csf=1&web=1&e=6N8BDm](url)

![TestScreenshot](https://github.com/user-attachments/assets/ab6c9141-a0bd-4a29-a5c7-3f5f48cb067b)





 ### Changelog

- Added SpandrelForce Class to the ETABS_oM Results namespace
- Added SpandrelForce value to Enum PierAndSpandrelResultType  in the Requests namespace
- Added calls to ETABS API functions to extract Spandrel Forces from ETABS Model in the ETABS Adapter

IsakNaslundBh reviewed (dismissed):
Have not tested, but just as a first comment, please have a look at the changes made by @peterjamesnugent in https://github.com/BHoM/ETABS_Toolkit/pull/455. Think the spandrel forces should follow suit, and also not inherit from BarForce

Chrisshort92 commented:
I have tested the functionality and happy that it works as expected with the test scripts provided. @GCRA101 I will wait until the comments above have been addressed before approving this PR.

GCRA101 commented:
@IsakNaslundBh , @Chrisshort92 ,
thank you for reviewing and for your comments.
I've updated the inheritance of the SpandrelForce Class to be from BarResult rather than from BarForce.

Updated code pushed to origin for your final check.


![SpandrelForceClassInheritance](https://github.com/user-attachments/assets/12e53235-96ab-4a18-a961-789f68833153)

peterjamesnugent commented on ETABS_oM/Requests/Enum/PierAndSpandrelResultType.cs:
```diff
@@ -27,7 +27,8 @@ namespace BH.oM.Adapters.ETABS.Requests
     public enum PierAndSpandrelResultType
     {
-        PierForce
+        PierForce,
+        SpandrelForce
```
I don't think this `Enum` is needed - see below comments on `BarResultRequest`.

GCRA101 replied:
Not 100% clear to me. Happy to discuss about it as soon as you're free. I understand this is not crucial for the code to work as expected. I might resolve this to allow for running BHoM Bot checks updating this asap in the next milestone.

peterjamesnugent reviewed (changes requested):
I couldn't get it to build without changes to the `SpandrelForce`, also it needs to update from origin/develop.

GCRA101 commented:
@peterjamesnugent thanks for your comments.
I've updated the code in line with them. Retested using the linked etabs model file and the gh script. All works fine.

GCRA101 reviewed (commented):
Thank for the comments. 
I've incorporated them in the code and tested it in Grasshopper.
Feature works as expected.

Chrisshort92 reviewed (commented):
Functionality has been tested and approved against the latest changes based on previous conversations about with @IsakNaslundBh and @peterjamesnugent.

GCRA101 commented on ETABS_oM/Results/SpandrelForce.cs:
```diff
@@ -0,0 +1,59 @@
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BH.oM.Base;
```
suggestion implemented and committed

GCRA101 commented on Etabs_Adapter/CRUD/Read/Results/PierAndSpandrelResults.cs:
```diff
@@ -105,6 +108,48 @@ private List<PierForce> GetPierForce(IList ids = null)
+                        position = 1;
+                    }
+
+                    SpandrelForce sf = new SpandrelForce(spandrelName[j],loadcaseNames[j], storyName[j], 0, 0, position, 2, p[j], v2[j], v3[j], t[j], m2[j], m3[j]);
```
Suggestion implemented and committed.

GCRA101 commented on ETABS_oM/Requests/PierAndSpandrelForceRequest.cs:
```diff
@@ -37,7 +37,7 @@ public class PierAndSpandrelForceRequest : IStructuralResultRequest
 
         [Description("Defines which type of results that should be extracted.")]
-        public virtual PierAndSpandrelResultType ResultType { get; set; } = PierAndSpandrelResultType.PierForce;
+        public virtual PierAndSpandrelResultType ResultType { get; set; } = new PierAndSpandrelResultType();
```
Suggestion implemented and committed.

peterjamesnugent replied:
Has this been added to the `BarResult` file I linked? You can use that to call it rather than the current implementation. I have not checked the test script but we want to have the same workflow for pulling `SpandrelForce` as we do all results.

Chrisshort92 reviewed (approved):
Functionality checked and code changes adopted. Approved for merge

Chrisshort92 commented:
@BHoMBot check required

GCRA101 commented:
@BHoMBot check copyright-compliance

GCRA101 commented:
@BHoMBot check ready-to-merge

peterjamesnugent reviewed (commented):
A few more changes need to be made to align it with the other ReadResults. Happy to run it through on a call tomorrow.

Chrisshort92 commented on ETABS_oM/Results/SpandrelForce.cs:
```diff
@@ -0,0 +1,99 @@
+            fz = fz;
+            mx = mx;
+            my = my;
+            mz = mz;
```
Comments adopted.

GCRA101 commented:
@peterjamesnugent , @Chrisshort92 

I've updated the initialization call of the PierAndSpandrel enum as suggested by Isak in today's call.
You can find the updated code for your review in the latest commit.
c01050f91e6fb78b7f248fef8b6158614154ee70

peterjamesnugent commented on Etabs_Adapter/Etabs_Adapter.csproj:
```diff
@@ -418,8 +421,13 @@
+
 </PostBuildEvent>
   </PropertyGroup>
+  <PropertyGroup>
+    <PreBuildEvent>
+    </PreBuildEvent>
+  </PropertyGroup>
```
Think this may have been added in error?

GCRA101 replied:
Hi @peterjamesnugent.
Not sure what happened with the code. Removed the four lines above.
Updated code in the latest commit

peterjamesnugent commented on Etabs_Adapter/Etabs_Adapter.csproj:
```diff
@@ -228,6 +228,9 @@
     <CodeAnalysisRuleSet>MinimumRecommendedRules.ruleset</CodeAnalysisRuleSet>
     <AssemblyName>ETABS18_Adapter</AssemblyName>
   </PropertyGroup>
+  <PropertyGroup>
+    <RunPostBuildEvent>Always</RunPostBuildEvent>
+  </PropertyGroup>
```
Added in error? Should not require any post build events.

GCRA101 replied:
Hi @peterjamesnugent ,
post build events calls above removed in latest commit 49033425304048adbfb3b9700a4f43c9bd11576c

peterjamesnugent reviewed (dismissed):
Some changes from csproj need to be removed.

Otherwise I am happy with the code changes - just @Chrisshort92 to review functionality.

Chrisshort92 reviewed (approved):
Functionality tested with most recent changes. Component working as expected.

GCRA101 commented:
@BHoMBot check copyright-compliance

GCRA101 commented:
@BHoMBot check null-handling

GCRA101 commented:
@BHoMBot check ready-to-merge
